StackOverflowError in DefaultBuilderProvider when a type has a public static generic method that returns the type

Hi !
I am playing a little with vavr (http://www.vavr.io) and mapstruct, but if I try to map beans with vavr lists, it fails with a stackoverflow error, even if I provide an implementation.

Mapper: 
```java
@Mapper
public interface PersonMapper {

    @Mappings(
            @Mapping(source = "age", target = "ageOfPerson")
    )
    PersonEntity toEntity(PersonDto dto);

    java.util.List<PersonEntity> toEntities(java.util.List<PersonDto> dtos);

    // Cause a stackoverlow. Why ?
    // List<PersonEntity> toVavrEntityList(List<PersonDto> dtos);

    // The same !

    // default List<PersonEntity> toVavrEntityList(List<PersonDto> dtos){
    //     return List.ofAll(toEntities(dtos.asJava()));
    // }

    @Mappings(
            @Mapping(source = "ageOfPerson", target = "age")
    )
    PersonDto toDto(PersonEntity entity);

    java.util.List<PersonDto> toDtos(java.util.List<PersonEntity> entities);

    default java.util.List<PersonDto> toDtosWithFancyRules(java.util.List<PersonEntity> entities) {
        return toDtos(entities).stream()
                .map(person -> person.toBuilder().age(person.getAge() * 2).build())
                .collect(Collectors.toList());
    }

}

```

Error:
```java
.../lightplayground/mapstruct/PersonMapper.java:11: error: Internal error in the mapping processor: java.la.StackOverflowError   
     at jdk.compiler/com.sun.tools.javac.util.Name.toString(Name.java:152)
     at jdk.compiler/com.sun.tools.javac.util.Name.charAt(Name.java:66) 
     at java.base/java.util.regex.Pattern$Slice.match(Pattern.java:4082)
     at java.base/java.util.regex.Pattern$Begin.match(Pattern.java:3656)
     at java.base/java.util.regex.Matcher.match(Matcher.java:1755)
     at java.base/java.util.regex.Matcher.matches(Matcher.java:713)     
     at org.mapstruct.ap.spi.DefaultBuilderProvider.shouldIgnore(DefaultBuilderProvider.java:281)
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuildMethods(DefaultBuilderProvider.java:229)   
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuilderInfo(DefaultBuilderProvider.java:167)    
     at org.mapstruct.ap.spi.DefaultBuilderProvider.findBuilderInfo(DefaultBuilderProvider.java:96)     
     at org.mapstruct.ap.internal.model.common.TypeFactory.findBuilder(TypeFactory.java:530)    
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:189)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:180)
     at org.mapstruct.ap.internal.model.common.BuilderType.create(BuilderType.java:98)  
     at org.mapstruct.ap.internal.model.common.Type.<init>(Type.java:160)
     at org.mapstruct.ap.internal.model.common.TypeFactory.getType(TypeFactory.java:280)
...
```


And beans are very simples. You can find a working example online:

Repo: 
https://github.com/remipassmoilesel/mapstruct-issue-03-2019

Mapper:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/blob/master/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java

Beans:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/tree/master/src/main/java/org/remipassmoilesel/lightplayground/mapstruct

Basic test:
https://github.com/remipassmoilesel/mapstruct-issue-03-2019/blob/master/src/test/java/org/remipassmoilesel/lightplayground/mapstruct/MapstructTest.java

Thanks for your help, and this awesome work on MapStruct !

Hi @filiphr, thanks for your help. I just tried master branch this morning, and it does not work, either with default implementation or not.

This repository can show you the issue: https://github.com/remipassmoilesel/mapstruct-issue-03-2019
(you will have to build mapstruct 1.4 SNAPSHOT locally before)

```
remipassmoilesel@XXXXXXXX: ~/projects/mapstruct-issue master   
$ 😝 ./gradlew clean test 

> Task :compileJava FAILED                                                                                                                                                                                         /home/remipassmoilesel/projects/mapstruct-issue/src/main/java/org/remipassmoilesel/lightplayground/mapstruct/PersonMapper.java:11: error: Internal error in the mapping processor: java.lang.StackOverflowError         at com.sun.tools.javac.code.Types$18.visitClassType(Types.java:2318)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$UnaryVisitor.visit(Types.java:4652)   at com.sun.tools.javac.code.Types.supertype(Types.java:2315)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1966)    at com.sun.tools.javac.code.Types$13.visitClassType(Types.java:1955)    at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.asSuper(Types.java:1952)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:921)      at com.sun.tools.javac.code.Types$4.visitClassType(Types.java:844)      at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:786)        at com.sun.tools.javac.code.Types$DefaultTypeVisitor.visit(Types.java:4571)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:840)     at com.sun.tools.javac.code.Types.isSubtype(Types.java:804)     at com.sun.tools.javac.code.Ty
....
```

--- a/processor/src/main/java/org/mapstruct/ap/spi/DefaultBuilderProvider.java
+++ b/processor/src/main/java/org/mapstruct/ap/spi/DefaultBuilderProvider.java
@@ -204,7 +204,14 @@ public class DefaultBuilderProvider implements BuilderProvider {
         return method.getParameters().isEmpty()
             && method.getModifiers().contains( Modifier.PUBLIC )
             && method.getModifiers().contains( Modifier.STATIC )
-            && !typeUtils.isSameType( method.getReturnType(), typeElement.asType() );
+            && method.getReturnType().getKind() != TypeKind.VOID
+            // Only compare raw elements
+            // Reason: if the method is a generic method (<T> Holder<T> build()) and the type element is (Holder<T>)
+            // then the return type of the method does not match the type of the type element
+            && !typeUtils.isSameType(
+            typeUtils.erasure( method.getReturnType() ),
+            typeUtils.erasure( typeElement.asType() )
+        );
     }
 
     /**

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Holder.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Holder.java
@@ -1,0 +1,35 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1751;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class Holder<T> {
+
+    private final T value;
+
+    public Holder(T value) {
+        this.value = value;
+    }
+
+    public T getValue() {
+        return value;
+    }
+
+    // If empty is considered as a builder creation method, this method would be the build method and would
+    // lead to a stackoverflow
+    @SuppressWarnings("unused")
+    public Holder<T> duplicate() {
+        return new Holder<>( value );
+    }
+
+    // This method should not be considered as builder creation method
+    @SuppressWarnings("unused")
+    public static <V> Holder<V> empty() {
+        return new Holder<>( null );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Issue1751Mapper.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Issue1751Mapper.java
@@ -1,0 +1,24 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1751;
+
+import org.mapstruct.Mapper;
+import org.mapstruct.factory.Mappers;
+
+/**
+ * @author Filip Hrisafov
+ */
+@Mapper
+public interface Issue1751Mapper {
+
+    Issue1751Mapper INSTANCE = Mappers.getMapper( Issue1751Mapper.class );
+
+    Target map(Source source);
+
+    default Holder<Target> mapToHolder(Source source) {
+        return new Holder<>( this.map( source ) );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Issue1751Test.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Issue1751Test.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1751;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mapstruct.ap.testutil.IssueKey;
+import org.mapstruct.ap.testutil.WithClasses;
+import org.mapstruct.ap.testutil.runner.AnnotationProcessorTestRunner;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+/**
+ * @author Filip Hrisafov
+ */
+@IssueKey("1772")
+@RunWith(AnnotationProcessorTestRunner.class)
+@WithClasses({
+    Holder.class,
+    Issue1751Mapper.class,
+    Source.class,
+    Target.class
+})
+public class Issue1751Test {
+
+    @Test
+    public void name() {
+        Source source = new Source();
+        source.setValue( "some value" );
+
+        Holder<Target> targetHolder = Issue1751Mapper.INSTANCE.mapToHolder( source );
+
+        assertThat( targetHolder.getValue() )
+            .extracting( Target::getValue )
+            .isEqualTo( "some value" );
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Source.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Source.java
@@ -1,0 +1,22 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1751;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class Source {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+}

--- a/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Target.java
+++ b/processor/src/test/java/org/mapstruct/ap/test/bugs/_1751/Target.java
@@ -1,0 +1,22 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1751;
+
+/**
+ * @author Filip Hrisafov
+ */
+public class Target {
+
+    private String value;
+
+    public String getValue() {
+        return value;
+    }
+
+    public void setValue(String value) {
+        this.value = value;
+    }
+}

